### ISSUE TITLE
Koe's Implementation Comments

### ISSUE THREAD

**Issue by UkoeHB** [seraphis-migration/wallet3#8]
Here are implementation comments/recommendations for using the seraphis library. I will periodically update it.

### Balance Recovery

- Balance recovery for legacy view-only wallets.
  - Reference (procedure required with helper tools): `legacy_ki_import_tool.h`.
  - The balance recovery procedure requires a 'key images only' pass where you collect key images with spent contexts from the blockchain and consume them in the enote store after importing recovered key images.
    - Reference: `sp::mocks::EnoteFindingContextLedgerMockLegacy` and `sp::mocks::ChunkConsumerMockLegacyIntermediate`.
  - Reference (mock implementation): unit test helper function `seraphis_enote_scanning: legacy_view_scan_recovery_cycle()`.
- Full balance recovery.
  - For the best results, always legacy full-scan first then seraphis full-scan (for view-only wallets, legacy full-scan means going through the entire balance recovery workflow).
  - To repair the enote store (`SpEnoteStore`) in case of an exception or other error during an update, save all of the last scanned block index from before the update (e.g. `SpEnoteStore::top_sp_scanned_block_index()`, then reset the enote store with them on failure (e.g. `SpEnoteStore::set_last_sp_scanned_index()`), and then re-scan to repair.
    - For efficiency, it's best to save the top scanned indices after each 'do scan' pass in the scan machine so automatically-recoverable failures will cause minimal rollbacks. However, index auto-rollbacks can also fail depending on the failure mode.
    - More aggressive recovery methods include: A) subtract 10-1000 blocks from the saved scan indices before resetting the enote store, B) discard the current enote store state and load from file if available, C) completely discard the current enote store and start over from refresh index (this may be necessary if the file gets corrupted, e.g. due to saving a corrupted enote store).

### Multisig

- Multisig nonce cache maintenance: a nonce cache's owner must manually clean up unused records since it's never obvious when a record can be safely discarded. It's possible for the normal multisig workflow to contain signature failures that don't remove records from the nonce cache (not to mention signing attempts aborted because one of the attempts succeeded, making the rest pointless). It is therefore necessary to periodically 'purge' the nonce cache of lingering values (only when the user/application is confident any nonces in the store aren't needed).
  - When an enote is spent on-chain and has an acceptable confirmation depth, any records associated with its onetime address can be removed. Nonce records of off-chain, unconfirmed, and inadequately-confirmed onchain spent enotes should not be purged, since there is still a chance they might be used.
  - A nonce record removed from the cache may still exist in persistent storage (a file somewhere). Users should ALWAYS refresh that storage after making a signature and before exposing that signature outside the local context, to avoid a situation where the signature is exported then the local context crashes/closes without updating the nonces in storage. Those sored nonces could be used to make another signature, thereby leaking the local signer's private multisig key material.
- Legacy key image recovery ceremony.
  - Reference: `multisig::mocks::mock_multisig_cn_key_image_recovery()` and unit test helper function `seraphis_multisig: refresh_user_enote_store_legacy_multisig()`.
  - The method `multisig::multisig_recover_cn_keyimage_cores()` returns information about failures encountered while assembling key images. It is up the to implementer whether to use this information (e.g. identify group members causing problems) or ignore it.
- Legacy -> Seraphis multisig account migration: this needs to A) update existing legacy accounts to permit aggregation-style signing (legacy accounts can't be discarded since multisigs may still have legacy funds; even if those funds are spent a reorg can always 'unspend' them), B) create Seraphis accounts using the same spend key as the legacy accounts.
  - Reference (legacy account update to support aggregation-style signing): unit test `multisig_signer_recommendations_recovery`.
  - Reference (make seraphis-era accounts from pre-existing legacy accounts): `multisig::mocks::mock_convert_multisig_accounts()` and unit test `multisig_account_conversions`.
  - For efficiency, do a legacy key image recovery ceremony in parallel with this migration ceremony. This way if all multisig members participate in migration and the group never makes legacy txs after migrating, then no more key image ceremonies will ever need to be performed.
  - WARNING: The currently implemented migration ceremony does not include a post-migration verification round (i.e. similar to the post-kex verification round that ensures other group members have completed their accounts). It is not implemented because when migrating accounts it's possible that not all signers are available. One workaround is to cache migration messages after completing your account migration, which can be relayed to other group members if they get blocked by a signer who stops cooperating.
- Validating a multisig tx proposal: must check all inputs are owned and spendable in the desired context (onchain, in the tx pool, etc.; spendable means unspent and unlocked), that the inputs will validate against the desired context (including that legacy ring members exist where they are supposed to and aren't locked), and that the multisig tx proposal is valid.
  - Reference: unit test helper function `seraphis_multisig: validate_multisig_tx_proposal()`.

### Future work

- `sp::mocks::MockLedgerContext`: this is doing the work of the daemon/tx pool plus find-received scanning blocks/pool cache.
- `sp::mocks::MockOffchainContext`: this isn't used right now, but represents an 'offchain context' that might be useful for protocols that use tx chaining (e.g. atomic swaps, multisig).
- `sp::SpEnoteStore` and `sp::SpEnoteStorePaymentValidator`: these are production-grade caches for managing owned enotes and interfacing with balance-recovery processes.
  - These are RAM-only caches that need to be integrated into a framework that involves loading to/from file. I strongly recommend keeping all file-related code out of these objects - use free functions to pipe data in and out.
- `sp::mocks::InputSelectorMockV1`: this is a naive inefficient input selector, a real input selector would likely involve many internal mechanisms and heuristics (e.g. filter for dust threshold to ignore, filter for origin statuses to permit, randomization and spread to mitigate timing analysis).
- `sp::LegacyDecoySelector (abstract)`: need a mature decoy selection algorithm (e.g. adapt the current gamma distribution approach).
- `sp::SpRefSetIndexMapper (abstract)`: need a mature decoy selection algorithm (e.g. adapt the current gamma distribution approach).
- `sp::TxValidationContext (abstract)`: needs to be implemented for real-world validation contexts (e.g. the daemon).
- `sp::ScanContextLedger (abstract)` and `sp::ScanContextNonLedger (abstract)`: need to be implemented for scanning the real ledger and the real tx pool respectively.
  - Notifications: the scan machine (see `scan_machine.h` and `scan_process_basic.h`) allows you to insert code between each 'do scan' state update (which involves scanning and consuming one chunk of data that can be arbitrarily small). Notifications about enote store changes (which are reported via `sp::EnoteStoreEvent`s) can be dispatched between those state updates.

### Miscellaneous comments

- Input selection: before selecting inputs, calculate the max number of inputs that can be added to your tx before it hits a size/weight limit. Max inputs = min(max inputs for size limit, max inputs by consensus).
- Making off-chain transactions:
  - Validating an off-chain tx should include checking the real ledger for key image duplicates (this step may not be required depending on the off-chain use-case; checking for key image duplicates is probably a race condition vs submitting the off-chain tx).
  - Spending legacy enotes: a partial transaction will contain full legacy membership proofs (CLSAGs) which must be validated against the real ledger by checking if the ring members (which are cached in {Ko, C} form) match on-chain legacy enotes at the specified indices.
- Legacy find-received scanning speedup (optional): for legacy view scanning using a remote node, reduce data transmission by...
    1) request {enote ephemeral pubkeys, view tags} from the remote node
    2) perform view tag checks on those partial enotes
    3) request full-enote information for view tag matches
    - This at least helps for bandwidth-limited scanning by reducing total data transmission by maybe 50-80%.
    - It may only be worthwhile to do this if the remote node is actually a remote find-received scanning service (doing seraphis find-received scanning), because the remote in that case will already be learning view-tag matches for seraphis enotes.
- When sending funds to yourself, make `jamtis::JamtisPaymentProposalSelfSendV1`s instead of `jamtis::JamtisPaymentProposalV1`s. This reduces the ability of remote scanners to track churns.
  - Test if a `jamtis::JamtisDestinationV1` is controlled by your keys with `jamtis::try_get_jamtis_index_from_destination_v1()`.
  - This workflow is not baked into the seraphis library's tx builder utilities in case an alternate wallet has other design goals.
- Balance tracking between tx submit and detecting it in the tx pool: you can load newly-created txs into an offchain ledger context and do balance recovery on that context. This allows you to propagate the balance changes of a tx to your wallet before a tx is detected in the tx pool. You can keep txs in the offchain context as long as they are needed. Removing a tx from the offchain context means subsequent refreshes using that context will clear corresponding information from the enote store (if the tx was added to the tx pool/ledger and you have already scanned to refresh your balance information, then rescanning an empty offchain context will not affect your enote store since origin/spent contexts prioritize statuses closer to on-chain).
- Legacy enote indexing: the seraphis library assumes everything is indexed uniformly, but legacy enotes have [different](https://libera.monerologs.net/monero-dev/20230411#c235649) indexing regimes. You can translate a ledger-global index into an amount-set index if you A) sort the non-zero amount sets by amount, B) build the ledger-global indexing by concatenating the sorted non-zero sets, C) append the 'ringct'-era set for zero amounts (plus ringct coinbase). To go  from ledger-global -> amount-set you need to know the global start index of the set your enote is in.
